File: 2.2_runround/runround.cpp
/*
ID: dermotl1
LANG: C++
PROG: runround
*/

#include <algorithm>
#include <fstream>
#include <iostream>
#include <vector>

using namespace std;

bool compute(int number) {
  vector<int> digits;
  int digit = 0;
  vector<bool> track(10, false);
  int num = number;

  while (num > 0) {
    digit = num % 10;
    if (!digit || track[digit]) return false;
    track[digit] = true;

    digits.push_back(digit);
    num /= 10;
  }

  reverse(digits.begin(), digits.end());

  vector<bool> saw(digits.size(), false);
  int watch = 0;      // where it is
  int digit_num = 0;  // how many numbers (it needs to be the number of digits
                      // because each is visited once)

  while (!saw[watch]) {
    saw[watch] = true;
    digit_num++;
    watch = (watch + digits[watch]) % digits.size();
  }
  if (digit_num == digits.size() && watch == 0) return true;
  return false;
}

int main() {
  ofstream fout("runround.out");
  ifstream fin("runround.in");

  int m;
  fin >> m;

  bool stop = false;
  int n = m + 1;

  while (!stop) {
    if (compute(n)) {
      stop = true;
      fout << n << "\n";
      return 0;
    }
    n++;
  }

  return 0;
}

File: 2.2_runround/runround.java
/*
ID: dermotl1
LANG: JAVA
PROG: runround
*/

import java.io.*;
import java.lang.*;
import java.util.*;

public class runround {
  public static boolean compute(int number) {
    ArrayList<Integer> digits = new ArrayList<>();
    int digit = 0;
    boolean[] track = new boolean[10];
    int num = number;

    while (num > 0) {
      digit = num % 10;
      if (digit == 0 || track[digit])
        return false;
      track[digit] = true;

      digits.add(digit);
      num /= 10;
    }

    Collections.reverse(digits);

    boolean[] saw = new boolean[digits.size()];
    int watch = 0;
    int digit_num = 0;

    while (!saw[watch]) {
      saw[watch] = true;
      digit_num++;
      watch = (watch + digits.get(watch)) % digits.size();
    }
    if (digit_num == digits.size() && watch == 0)
      return true;
    return false;
  }

  public static void main(String[] args) throws FileNotFoundException {
    Scanner in = new Scanner(new File("runround.in"));
    PrintWriter out = new PrintWriter(new File("runround.out"));

    int m;
    m = in.nextInt();

    boolean stop = false;
    int n = m + 1;

    while (!stop) {
      if (compute(n)) {
        stop = true;
        out.println(n);
        in.close();
        out.close();
      }
      n++;
    }

    in.close();
    out.close();
  }
}
